### PR TITLE
feat(renderer): add SearchTermParser for filtered search inputs 

### PR DESCRIPTION
**Pull request by simonrey1**
### What does this PR do?

Introduces a generic `SearchTermParser` class that parses a search string into plain-text terms and typed filter values, respecting double-quoted multi-word values (e.g. `category:"Extension Packs"`).

The class:
- Takes allowed filter prefixes in its constructor (e.g. `category`, `keyword`, `is`, `not`)
- Normalizes input to lowercase
- Exposes `terms: string[]` for plain-text search words
- Provides a type-safe `getFilter(name): string[]` method constrained to the declared filter prefixes via generics (`SearchTermParser<F extends string>`)

This is extracted as a standalone building block so it can be validated independently. The integration into the extension catalog filters is available for review in the draft PR #16993.

### Screenshot / video of UI

N/A — no UI changes, this adds a utility class with tests only.

### What issues does this PR fix or reference?

Part of #16971

### How to test this PR?

```bash
pnpm exec vitest run packages/renderer/src/stores/search-term-parser.spec.ts
```

- [x] Tests are covering the bug fix or the new feature

AI-assisted: cursor

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: packages/renderer/src/lib/search/search-term-parser.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
+/**********************************************************************
```
Comment (axel7083) [inline packages/renderer/src/lib/search/search-term-parser.ts:1]: wrong location? we should avoid having logic related code in the stores folder

File: packages/renderer/src/lib/search/search-term-parser.spec.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+import { describe, expect, test } from 'vitest';
+
+import { SearchTermParser } from './search-term-parser';
+
+describe('SearchTermParser', () => {
```
Comment (axel7083) [inline packages/renderer/src/lib/search/search-term-parser.spec.ts:23]: suggestion(non-blocking): you could simplify the code by using `test.each`

File: packages/renderer/src/lib/search/search-term-parser.ts (hunk 1 above)
Comment (simonrey1) [reply packages/renderer/src/lib/search/search-term-parser.ts:1]: It's alongside the search-utils.ts where the original logic is, but I will find another place, good remark!

Comment (simonrey1) [reply packages/renderer/src/lib/search/search-term-parser.ts:1]: Proposition: I moved it to `lib/search`

Let me know if it's better

Comment (axel7083) [reply packages/renderer/src/lib/search/search-term-parser.ts:1]: yes I guess

File: packages/renderer/src/lib/search/search-term-parser.spec.ts (hunk 2 above)
Comment (simonrey1) [reply packages/renderer/src/lib/search/search-term-parser.spec.ts:23]: Great remark! I was caught for not doing it while always asking others. I have pushed a fix.

File: packages/renderer/src/lib/search/search-term-parser.spec.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  const FILTERS = ['category', 'keyword', 'is', 'not'] as const;
+  type Filter = (typeof FILTERS)[number];
+
+  test.each<{
+    input: string;
+    terms: string[];
+    filters?: Partial<Record<Filter, string[]>>;
```
Comment (axel7083) [inline packages/renderer/src/lib/search/search-term-parser.spec.ts]: question(blocking): what is the point of `Partial`?

File: packages/renderer/src/lib/search/search-term-parser.spec.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+import { describe, expect, test } from 'vitest';
+
+import { SearchTermParser } from './search-term-parser';
+
+describe('SearchTermParser', () => {
+  const FILTERS = ['category', 'keyword', 'is', 'not'] as const;
+  type Filter = (typeof FILTERS)[number];
```
Comment (axel7083) [inline packages/renderer/src/lib/search/search-term-parser.spec.ts:25]: Why do you need to define them before, just use string?

**Review by axel7083 — APPROVED** [review]

File: packages/renderer/src/lib/search/search-term-parser.spec.ts (hunk 3 above)
Comment (simonrey1) [reply packages/renderer/src/lib/search/search-term-parser.spec.ts]: Nothing specific, I removed it

File: packages/renderer/src/lib/search/search-term-parser.spec.ts (hunk 4 above)
Comment (simonrey1) [reply packages/renderer/src/lib/search/search-term-parser.spec.ts:25]: As discussed in messages, you seem ok with this approach now

**Review by benoitf — COMMENTED** [review]
question: do you think we could use an existing npmjs library to handle this rather than building our own parser ?

it could allow to handle "and", "or", "not expression maybe

**Comment by benoitf** [comment]
well, I don't see why in lib/search we would hardcode thinks like 
```ts
const FILTERS = ['category', 'keyword', 'is', 'not'] as const;
```

category is only meaningful for the extension page, for containers it would be another filter.

> for instance, it seems that these are not adopted and to begin with, I think it's pretty good to centralize the logic here with this pretty simple regex. 

To me centralize a logic means that it works for all use-cases but here it's only useful for extensions.

This is why either library is generic, relying on stuff that can be really extended/used by all search pages, or it's not generic meaning it can't go to lib/search

**Comment by simonrey1** [comment]
> well, I don't see why in lib/search we would hardcode thinks like
> 
> ```ts
> const FILTERS = ['category', 'keyword', 'is', 'not'] as const;
> ```
> 
> category is only meaningful for the extension page, for containers it would be another filter.
> 
> > for instance, it seems that these are not adopted and to begin with, I think it's pretty good to centralize the logic here with this pretty simple regex.
> 
> To me centralize a logic means that it works for all use-cases but here it's only useful for extensions.
> 
> This is why either library is generic, relying on stuff that can be really extended/used by all search pages, or it's not generic meaning it can't go to lib/search

It is only present in the test.

**Comment by benoitf** [comment]
please use https://www.npmjs.com/package/search-string as it would also handle for example single quotes or other corner cases

**Comment by simonrey1** [comment]
Moving to draft, waiting for https://github.com/podman-desktop/podman-desktop/pull/17026

**Comment by simonrey1** [comment]
Closing/Reopening as checks seemed stuck

File: packages/renderer/src/lib/search/search-term-parser.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      if (!segment.negated && segment.text.length > 0) {
+        terms.push(segment.text);
+      }
+    }
+
+    for (const condition of parsed.getConditionArray()) {
+      if (condition.negated) continue;
+      const key = condition.keyword as F;
+      if (allowedSet.has(key)) {
+        filters.get(key)!.push(condition.value);
```
Comment (benoitf) [inline packages/renderer/src/lib/search/search-term-parser.ts]: issue: do not use ! Symbol

**Comment by simonrey1** [comment]
Ready for review

**Review by axel7083 — APPROVED** [review]

**Review by benoitf — APPROVED** [review]
